### ISSUE TITLE
Exception in thread EnqueueThread QueueInput/input_queue - msgpack exceeds max_bin_len()

### ISSUE THREAD

**Issue by ngatran00** [tensorpack/tensorpack#1191]
### 1. What you did:

(1) **If you're using examples, what's the command you run:**
`python3 train.py --config MODE_MASK=True MODE_FPN=True DATA.BASEDIR="coco" BACKBONE.WEIGHTS=R50.npz`

(2) **If you're using examples, have you made any changes to the examples? Paste `git status; git diff` here:**
Changing config to adapt to widerface dataset

### 2. What you observed:

(1) **Include the ENTIRE logs here:**
```
[32m[0515 23:27:44 @base.py:275][0m Start Epoch 15 ...
[32m[0515 23:28:41 @input_source.py:173][0m [4m[5m[31mERR[0m [EnqueueThread] Exception in thread EnqueueThread QueueInput/input_queue:
Traceback (most recent call last):
  File "/home/nga/FasterRCNN/tensorpack/input_source/input_source.py", line 161, in run
    dp = next(self._itr)
  File "/home/nga/FasterRCNN/tensorpack/dataflow/common.py", line 370, in __iter__
    for dp in self.ds:
  File "/home/nga/FasterRCNN/tensorpack/dataflow/parallel_map.py", line 314, in __iter__
    for dp in super(MultiProcessMapDataZMQ, self).__iter__():
  File "/home/nga/FasterRCNN/tensorpack/dataflow/parallel_map.py", line 89, in __iter__
    for dp in self.get_data_non_strict():
  File "/home/nga/FasterRCNN/tensorpack/dataflow/parallel_map.py", line 65, in get_data_non_strict
    ret = self._recv()
  File "/home/nga/FasterRCNN/tensorpack/dataflow/parallel_map.py", line 309, in _recv
    dp = loads(msg[1])
  File "/home/nga/FasterRCNN/tensorpack/utils/serialize.py", line 43, in loads_msgpack
    max_str_len=MAX_MSGPACK_LEN)
  File "/home/nga/.local/lib/python3.6/site-packages/msgpack_numpy.py", line 214, in unpackb
    return _unpackb(packed, **kwargs)
  File "msgpack/_unpacker.pyx", line 200, in msgpack._unpacker.unpackb
ValueError: 1916481600 exceeds max_bin_len(1000000000)
[32m[0515 23:28:41 @input_source.py:179][0m [EnqueueThread] Thread EnqueueThread QueueInput/input_queue Exited.
[32m[0515 23:28:58 @base.py:291][0m Training was stopped by exception FIFOQueue '_0_QueueInput/input_queue' is closed and has insufficient elements (requested 1, current size 0)
	 [[node QueueInput/input_deque (defined at /home/nga/FasterRCNN/tensorpack/input_source/input_source.py:272)  = QueueDequeueV2[component_types=[DT_FLOAT, DT_INT32, DT_FLOAT, DT_INT32, DT_FLOAT, DT_INT32, DT_FLOAT, DT_INT32, DT_FLOAT, DT_INT32, DT_FLOAT, DT_FLOAT, DT_INT64, DT_UINT8], timeout_ms=-1, _device="/job:localhost/replica:0/task:0/device:CPU:0"](QueueInput/input_queue)]]

Caused by op 'QueueInput/input_deque', defined at:
  File "train.py", line 120, in <module>
    launch_train_with_config(traincfg, trainer)
  File "/home/nga/FasterRCNN/tensorpack/train/interface.py", line 91, in launch_train_with_config
    model.build_graph, model.get_optimizer)
  File "/home/nga/FasterRCNN/tensorpack/utils/argtools.py", line 176, in wrapper
    return func(*args, **kwargs)
  File "/home/nga/FasterRCNN/tensorpack/train/tower.py", line 224, in setup_graph
    train_callbacks = self._setup_graph(input, get_cost_fn, get_opt_fn)
  File "/home/nga/FasterRCNN/tensorpack/train/trainers.py", line 189, in _setup_graph
    grad_list = self._builder.call_for_each_tower(tower_fn)
  File "/home/nga/FasterRCNN/tensorpack/graph_builder/training.py", line 226, in call_for_each_tower
    use_vs=[False] + [True] * (len(self.towers) - 1))
  File "/home/nga/FasterRCNN/tensorpack/graph_builder/training.py", line 122, in build_on_towers
    return DataParallelBuilder.call_for_each_tower(*args, **kwargs)
  File "/home/nga/FasterRCNN/tensorpack/graph_builder/training.py", line 117, in call_for_each_tower
    ret.append(func())
  File "/home/nga/FasterRCNN/tensorpack/train/tower.py", line 252, in get_grad_fn
    inputs = input.get_input_tensors()
  File "/home/nga/FasterRCNN/tensorpack/input_source/input_source_base.py", line 83, in get_input_tensors
    return self._get_input_tensors()
  File "/home/nga/FasterRCNN/tensorpack/input_source/input_source.py", line 272, in _get_input_tensors
    ret = self.queue.dequeue(name='input_deque')
  File "/home/nga/.local/lib/python3.6/site-packages/tensorflow/python/ops/data_flow_ops.py", line 435, in dequeue
    self._queue_ref, self._dtypes, name=name)
  File "/home/nga/.local/lib/python3.6/site-packages/tensorflow/python/ops/gen_data_flow_ops.py", line 3741, in queue_dequeue_v2
    timeout_ms=timeout_ms, name=name)
  File "/home/nga/.local/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 787, in _apply_op_helper
    op_def=op_def)
  File "/home/nga/.local/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py", line 488, in new_func
    return func(*args, **kwargs)
  File "/home/nga/.local/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3274, in create_op
    op_def=op_def)
  File "/home/nga/.local/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1770, in __init__
    self._traceback = tf_stack.extract_stack()

OutOfRangeError (see above for traceback): FIFOQueue '_0_QueueInput/input_queue' is closed and has insufficient elements (requested 1, current size 0)
	 [[node QueueInput/input_deque (defined at /home/nga/FasterRCNN/tensorpack/input_source/input_source.py:272)  = QueueDequeueV2[component_types=[DT_FLOAT, DT_INT32, DT_FLOAT, DT_INT32, DT_FLOAT, DT_INT32, DT_FLOAT, DT_INT32, DT_FLOAT, DT_INT32, DT_FLOAT, DT_FLOAT, DT_INT64, DT_UINT8], timeout_ms=-1, _device="/job:localhost/replica:0/task:0/device:CPU:0"](QueueInput/input_queue)]]

```

(2) **Other observations, if any:**
The error occurs at random epoch, sometimes 12, sometimes 6, 9 and so on.

### 3. What you expected, if not obvious.
How can I resolve this issue?

### 4. Your environment:
```
Python==3.6.7
Tensorpack==0.9.4
TensorFlow==1.12.0
msgpack==0.5.6
```
1 GPU - GeForce GTX 1080
Free RAM              124.25/125.82 GB

**Comment by ppwwyyxx** [comment]
> ValueError: 1916481600 exceeds max_bin_len(1000000000)


This indicates that the size of data for __one image__,  after the preprocessing, is around 2GB.
The data only contains the image and some annotations and should not have this size. This indicates you may have some erroneous data format.

**Comment by ngatran00** [comment]
Thank you for the quick reply!

I didn't modify anything in the data.py. In addition, I have checked the data input and its size, and the images seem to be fine.

The error occurs for different images also. Could it be the number of masks each image has? There are images which have over 80 masks.

**Comment by ppwwyyxx** [comment]
80 is a small number.
Could you post your changes and your entire logs, as the issue template requested?

**Comment by ngatran00** [comment]
Difference in config.py
```
88c88
< _C.DATA.TRAIN = ('WIDER_train',)   # i.e. trainval35k, AKA train2017
---
> _C.DATA.TRAIN = ('coco_train2014', 'coco_valminusminival2014')   # i.e. trainval35k, AKA train2017
90c90
< _C.DATA.VAL = ('WIDER_val', )  # AKA val2017
---
> _C.DATA.VAL = ('coco_minival2014', )  # AKA val2017
92c92
< _C.DATA.NUM_CATEGORY = 1  # without the background class (e.g., 80 for COCO)
---
> _C.DATA.NUM_CATEGORY = 80  # without the background class (e.g., 80 for COCO)
```

Difference in coco.py
```
28c28
<     COCO_id_to_category_id = {0:1}  # noqa
---
>     COCO_id_to_category_id = {13: 12, 14: 13, 15: 14, 16: 15, 17: 16, 18: 17, 19: 18, 20: 19, 21: 20, 22: 21, 23: 22, 24: 23, 25: 24, 27: 25, 28: 26, 31: 27, 32: 28, 33: 29, 34: 30, 35: 31, 36: 32, 37: 33, 38: 34, 39: 35, 40: 36, 41: 37, 42: 38, 43: 39, 44: 40, 46: 41, 47: 42, 48: 43, 49: 44, 50: 45, 51: 46, 52: 47, 53: 48, 54: 49, 55: 50, 56: 51, 57: 52, 58: 53, 59: 54, 60: 55, 61: 56, 62: 57, 63: 58, 64: 59, 65: 60, 67: 61, 70: 62, 72: 63, 73: 64, 74: 65, 75: 66, 76: 67, 77: 68, 78: 69, 79: 70, 80: 71, 81: 72, 82: 73, 84: 74, 85: 75, 86: 76, 87: 77, 88: 78, 89: 79, 90: 80}  # noqa
34c34,35
<     class_names = ["face"]  # noqa
---
>     class_names = [
>         "person", "bicycle", "car", "motorcycle", "airplane", "bus", "train", "truck", "boat", "traffic light", "fire hydrant", "stop sign", "parking meter", "bench", "bird", "cat", "dog", "horse", "sheep", "cow", "elephant", "bear", "zebra", "giraffe", "backpack", "umbrella", "handbag", "tie", "suitcase", "frisbee", "skis", "snowboard", "sports ball", "kite", "baseball bat", "baseball glove", "skateboard", "surfboard", "tennis racket", "bottle", "wine glass", "cup", "fork", "knife", "spoon", "bowl", "banana", "apple", "sandwich", "orange", "broccoli", "carrot", "hot dog", "pizza", "donut", "cake", "chair", "couch", "potted plant", "bed", "dining table", "toilet", "tv", "laptop", "mouse", "remote", "keyboard", "cell phone", "microwave", "oven", "toaster", "sink", "refrigerator", "book", "clock", "vase", "scissors", "teddy bear", "hair drier", "toothbrush"]  # noqa
213,214c214,216
<     for split in ["WIDER_train", "WIDER_val"]:
<         DatasetRegistry.register(split, lambda x=split: COCODetection(basedir, x))
---
>     for split in ["train2017", "val2017", "train2014", "val2014",
>                   "valminusminival2014", "minival2014"]:
>         DatasetRegistry.register("coco_" + split, lambda x=split: COCODetection(basedir, x))
218,219c220,221
<     basedir = 'coco'
<     c = COCODetection(basedir, 'WIDER_train')
---
>     basedir = '~/data/coco'
>     c = COCODetection(basedir, 'train2014')
```

The log file is attached below
[log.log](https://github.com/tensorpack/tensorpack/files/3188564/log.log)

Thank you!

**Comment by ppwwyyxx** [comment]
The log and config looks correct. I think there is still something wrong in your dataset. You can try the following:
1. Run `python data.py` directly (after removing the path in the end of `data.py`). This will run the dataloader alone without training, and should throw the same error.
2. If it does throw the error. Add:
```python
from tensorpack.utils import dumps
print("size:", len(dumps(ret)))
```
in the end of the preprocessor, in data.py, and set `DATA.NUM_WORKERS=0`. 
This will tell you the size of the data, and if it is too large, you can dump the data to a file to further understand what went wrong.

**Comment by ppwwyyxx** [comment]
A quick shot is to `print(ret['gt_masks'].shape)` to see whether it's reasonable. This is most likely what's taking the space.

**Comment by ngatran00** [comment]
Sorry for the late reply.

Once I set `DATA.NUM_WORKERS=0`, the training works fine and there is no more error.

Thank you so much!

**Comment by ppwwyyxx** [comment]
setting `DATA.NUM_WORKERS=0` will skip the codepath that throws the error and does not fix your issue. It will also make your training slower.

**Comment by ngatran00** [comment]
So I ran the dataloader alone, and the script crashed at this image 
![16_Award_Ceremony_Awards_Ceremony_16_411](https://user-images.githubusercontent.com/33063754/58117170-43bed400-7bfe-11e9-9b11-ff38dd8add0b.jpg)


In the terminal, the size and mask shape of the image is `size: 36678475 mask shape (17, 800, 1192)`, but it still prints out the error `ValueError: 1675200000 exceeds max_bin_len(1000000000)`

The data size and the mask shape seems reasonable. Do you have any insight into this?

**Comment by ppwwyyxx** [comment]
When you run it with NUM_WORKERS>1, it may not print the correct information which leads to the crash. 
Could you use NUM_WORKERS=0 instead, and see if there is any data that has extremely large data size (by `len(dumps(x))`?

**Comment by ngatran00** [comment]
How large do you mean by extremely large data size? 

I ran the script again, and there is an image with the original size of 205,135 bytes, but after preprocessing, it prints `size: 158703939 mask shape (170, 800, 1028)`. Is that extremely large?

**Comment by ppwwyyxx** [comment]
This still looks like a reasonable size (158M). I was hoping you could find anything larger than 1G, e.g. 1675200000. 
Theoretically if there are data > 1G in when you use >1 workers, there should be something > 1G when you use 0 workers as well. Otherwise this is very strange. If you can't find anything>1G could you try upgrade msgpack and try again?

**Comment by ngatran00** [comment]
There is a few images with a lot of masks, for example an image with `size: 1151759322 mask shape (1139, 800, 1241)`. In this case, should I remove some of the masks, or how should I approach it?

Thank you!

**Comment by ppwwyyxx** [comment]
That's interesting. Do you really have images with >1000 instance annotations?

**Comment by ngatran00** [comment]
The image I mentioned is this one
![12_Group_Large_Group_12_Group_Large_Group_12_31](https://user-images.githubusercontent.com/33063754/58190745-74b20e00-7cbd-11e9-9f0d-0c056f4c961d.jpg)

I suppose it could have over 1000 annotations, since it is face detection

**Comment by ppwwyyxx** [comment]
OMG.
Then this is a real issue. For now you can unblock yourself by changing the size limit at:
https://github.com/tensorpack/tensorpack/blob/42416945c1e36a5f1d4350ee9c1ae8b134cbe841/tensorpack/utils/serialize.py#L19

I'll see what's a better way to fix this.

**Comment by ngatran00** [comment]
Oh okay, thank you so much 😄

**Comment by ppwwyyxx** [comment]
https://github.com/tensorpack/tensorpack/commit/0bbec1e466b78794dbd22cbd70770cc55779e9fd packs the masks into bit array and should reduce the size of data by 8x. I'll run some more tests before merging it.

**Comment by XuanBaby** [comment]
@ppwwyyxx  hi,I also meet the similiar error,could you give me some advice?
`W tensorflow/core/kernels/queue_base.cc:277] _0_QueueInput/input_queue: Skipping cancelled enqueue attempt with queue not closed
Traceback (most recent call last):
  File "/public/home/yangxuan1/Build/python-3.6.8/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1356, in _do_call
    return fn(*args)
  File "/public/home/yangxuan1/Build/python-3.6.8/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1341, in _run_fn
    options, feed_dict, fetch_list, target_list, run_metadata)
  File "/public/home/yangxuan1/Build/python-3.6.8/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1429, in _call_tf_sessionrun
    run_metadata)
tensorflow.python.framework.errors_impl.InvalidArgumentError: 2 root error(s) found.
  (0) Invalid argument: Input to reshape is a tensor with 600 values, but the requested shape has 120000
         [[{{node gradients/rpn_1/Reshape_grad/Reshape}}]]
  (1) Invalid argument: Input to reshape is a tensor with 600 values, but the requested shape has 120000
         [[{{node gradients/rpn_1/Reshape_grad/Reshape}}]]
         [[gradients/AddN_154/_4309]]
0 successful operations.
0 derived errors ignored. `

`During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "train.py", line 130, in <module>
    launch_train_with_config(traincfg, trainer)
  File "../../tensorpack/train/interface.py", line 101, in launch_train_with_config
    extra_callbacks=config.extra_callbacks)
  File "../../tensorpack/train/base.py", line 342, in train_with_defaults
    steps_per_epoch, starting_epoch, max_epoch)
  File "../../tensorpack/train/base.py", line 314, in train
    self.main_loop(steps_per_epoch, starting_epoch, max_epoch)
  File "../../tensorpack/utils/argtools.py", line 168, in wrapper
    return func(*args, **kwargs)
  File "../../tensorpack/train/base.py", line 279, in main_loop
    self.run_step()  # implemented by subclass
  File "../../tensorpack/train/base.py", line 179, in run_step
    self.hooked_sess.run(self.train_op)
  File "/public/home/yangxuan1/Build/python-3.6.8/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 754, in run
    run_metadata=run_metadata)
  File "/public/home/yangxuan1/Build/python-3.6.8/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 1252, in run
    run_metadata=run_metadata)
  File "/public/home/yangxuan1/Build/python-3.6.8/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 1353, in run
    raise six.reraise(*original_exc_info)
  File "/public/home/yangxuan1/Build/python-3.6.8/lib/python3.6/site-packages/six.py", line 693, in reraise
    raise value
  File "/public/home/yangxuan1/Build/python-3.6.8/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 1338, in run
    return self._sess.run(*args, **kwargs)
  File "/public/home/yangxuan1/Build/python-3.6.8/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 1411, in run
    run_metadata=run_metadata)
  File "/public/home/yangxuan1/Build/python-3.6.8/lib/python3.6/site-packages/tensorflow/python/training/monitored_session.py", line 1169, in run
    return self._sess.run(*args, **kwargs)
  File "/public/home/yangxuan1/Build/python-3.6.8/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 950, in run
    run_metadata_ptr)
  File "/public/home/yangxuan1/Build/python-3.6.8/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1173, in _run
    feed_dict_tensor, options, run_metadata)
  File "/public/home/yangxuan1/Build/python-3.6.8/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1350, in _do_run
    run_metadata)
  File "/public/home/yangxuan1/Build/python-3.6.8/lib/python3.6/site-packages/tensorflow/python/client/session.py", line 1370, in _do_call
    raise type(e)(node_def, op, message)
tensorflow.python.framework.errors_impl.InvalidArgumentError: 2 root error(s) found.`

`Original stack trace for 'gradients/rpn_1/Reshape_grad/Reshape':
  File "train.py", line 130, in <module>
    launch_train_with_config(traincfg, trainer)
  File "../../tensorpack/train/interface.py", line 91, in launch_train_with_config
    model.build_graph, model.get_optimizer)
  File "../../tensorpack/utils/argtools.py", line 168, in wrapper
    return func(*args, **kwargs)
  File "../../tensorpack/train/tower.py", line 224, in setup_graph
    train_callbacks = self._setup_graph(input, get_cost_fn, get_opt_fn)
  File "../../tensorpack/train/trainers.py", line 447, in _setup_graph
    grads = self._make_get_grad_fn(input, get_cost_fn, get_opt_fn)()
  File "../../tensorpack/train/tower.py", line 281, in get_grad_fn
    return compute_grad_from_inputs(*inputs)
  File "../../tensorpack/train/tower.py", line 276, in compute_grad_from_inputs
    aggregation_method=self.AGGREGATION_METHOD)
  File "../../tensorpack/tfutils/optimizer.py", line 29, in compute_gradients
    return self._opt.compute_gradients(*args, **kwargs)
  File "/public/home/yangxuan1/Build/python-3.6.8/lib/python3.6/site-packages/tensorflow/python/training/optimizer.py", line 512, in compute_gradients
    colocate_gradients_with_ops=colocate_gradients_with_ops)
  File "/public/home/yangxuan1/Build/python-3.6.8/lib/python3.6/site-packages/tensorflow/python/ops/gradients_impl.py", line 158, in gradients
    unconnected_gradients)
  File "/public/home/yangxuan1/Build/python-3.6.8/lib/python3.6/site-packages/tensorflow/python/ops/gradients_util.py", line 731, in _GradientsHelper
    lambda: grad_fn(op, *out_grads))
  File "/public/home/yangxuan1/Build/python-3.6.8/lib/python3.6/site-packages/tensorflow/python/ops/gradients_util.py", line 403, in _MaybeCompile
    return grad_fn()  # Exit early
  File "/public/home/yangxuan1/Build/python-3.6.8/lib/python3.6/site-packages/tensorflow/python/ops/gradients_util.py", line 731, in <lambda>
    lambda: grad_fn(op, *out_grads))
  File "/public/home/yangxuan1/Build/python-3.6.8/lib/python3.6/site-packages/tensorflow/python/ops/array_grad.py", line 540, in _ReshapeGrad
    return [array_ops.reshape(grad, array_ops.shape(op.inputs[0])), None]
  File "/public/home/yangxuan1/Build/python-3.6.8/lib/python3.6/site-packages/tensorflow/python/ops/gen_array_ops.py", line 7715, in reshape
    "Reshape", tensor=tensor, shape=shape, name=name)
  File "/public/home/yangxuan1/Build/python-3.6.8/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 788, in _apply_op_helper
    op_def=op_def)
  File "/public/home/yangxuan1/Build/python-3.6.8/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py", line 507, in new_func
    return func(*args, **kwargs)
  File "/public/home/yangxuan1/Build/python-3.6.8/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3616, in create_op
    op_def=op_def)
  File "/public/home/yangxuan1/Build/python-3.6.8/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 2005, in __init__
    self._traceback = tf_stack.extract_stack()`

**Comment by ppwwyyxx** [comment]
For anyone to better understand and diagnose your issue, please post relevant details following the issue template (click "New Issue" -> "Unexpected Problems / Bugs", or visit [this link to post issues about unexpected problems](https://github.com/tensorpack/tensorpack/issues/new?template=unexpected-problems---bugs.md))

**Comment by XuanBaby** [comment]
@ppwwyyxx  ok I have ask question as [https://github.com/tensorpack/tensorpack/issues/1373](url)
